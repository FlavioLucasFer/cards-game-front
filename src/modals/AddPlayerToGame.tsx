import { useState } from 'react';
import Button from 'react-bootstrap/Button';
import Modal, { ModalProps } from 'react-bootstrap/Modal';
import axiosInstance from '../api';
import LabeledInput from '../components/Input';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { Player, setPlayer } from '../redux/reducers/player';
import axios from 'axios';

const propTypes = {
    gameUuid: PropTypes.string.isRequired,
    onJoin: PropTypes.func.isRequired,
};

type AddPlayerToGameProps = ModalProps & PropTypes.InferProps<typeof propTypes>

const AddPlayerToGame: React.FC<AddPlayerToGameProps> = props => {
    const {
        gameUuid,
        onJoin,
        ...modalProps
    } = props;

    const dispatch = useDispatch();
    const [nickname, setNickname] = useState<string>('');

    const handleJoin = async () => {
        try {
            const { data } = await axiosInstance.post<Player>(
                `/games/${gameUuid}/players`, 
                { nickname },
            );

            dispatch(setPlayer(data));
            
            onJoin();
        } catch (err) {
            if (axios.isAxiosError(err))
                console.log('err:', err.response);
        }
    }

    const handleNicknameChange = (e: React.FormEvent<HTMLInputElement>) => {
        setNickname(e.currentTarget.value);
    }

    return (
        <Modal
            {...modalProps}
            aria-labelledby="contained-modal-title-vcenter"
            centered
        >
            <Modal.Body>
                <LabeledInput 
                    id='nickname-input'
                    label='Nickname'
                    value={nickname}
                    onChange={handleNicknameChange}
                    autoFocus
                />
            </Modal.Body>
            <Modal.Footer>
                <Button 
                    onClick={handleJoin}
                    disabled={!nickname}
                >
                    Join
                </Button>
            </Modal.Footer>
        </Modal>
    );
}

AddPlayerToGame.propTypes = propTypes;

export default AddPlayerToGame;
