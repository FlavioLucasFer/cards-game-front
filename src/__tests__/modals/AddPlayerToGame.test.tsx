import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import AddPlayerToGame from '../../modals/AddPlayerToGame';

configure({ adapter: new Adapter() });

const mockedHandleJoin = jest.fn();
const mockedUseDispatch = jest.fn();

jest.mock('react-redux', () => ({
    ...jest.requireActual('react-redux') as any,
    useDispatch: () => mockedUseDispatch,
}));

describe('AddPlayerToGame tests', () => {
    it('should render properly', () => {
        const wrapper = shallow(
            <AddPlayerToGame 
                gameUuid='680fbf96-bf45-4ae5-8cce-a3d6d6e876bb'
                onJoin={mockedHandleJoin}
            />
        );

        expect(wrapper.debug()).toMatchSnapshot();
    });
});
