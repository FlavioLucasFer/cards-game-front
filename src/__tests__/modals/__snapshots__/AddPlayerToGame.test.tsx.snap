// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddPlayerToGame tests should render properly 1`] = `
"<Modal aria-labelledby=\\"contained-modal-title-vcenter\\" centered={true} show={false} backdrop={true} keyboard={true} autoFocus={true} enforceFocus={true} restoreFocus={true} animation={true} dialogAs={{...}}>
  <ModalBody>
    <Input id=\\"nickname-input\\" label=\\"Nickname\\" value=\\"\\" onChange={[Function: handleNicknameChange]} autoFocus={true} />
  </ModalBody>
  <ModalFooter>
    <Button onClick={[Function: handleJoin]} disabled={true} variant=\\"primary\\" active={false}>
      Join
    </Button>
  </ModalFooter>
</Modal>"
`;
